src/Stratis.SmartContracts.CLR.Tests/SmartContracts/SystemContractsDictionary.cs

```diff
@@ -0,0 +1,205 @@
+﻿using Stratis.SmartContracts;
+using ECRecover=Stratis.SCL.Crypto.ECRecover;
+
+public struct WhiteListEntry
+{
+    public UInt256 CodeHash;
+    public Address LastAddress;
+    public string Name;
+}
+
+public class SystemContractsDictionary : SmartContract
+{
+    const string primaryGroup = "main";
+
+    public SystemContractsDictionary(ISmartContractState state) : base(state)
+    {
+        this.State.SetArray($"Signatories:{primaryGroup}", new[] { new Address(0, 0, 0, 0, 0), new Address(0, 0, 0, 0, 1), new Address(0, 0, 0, 0, 2) });
+        this.State.SetUInt32($"Quorum:{primaryGroup}", 2);
+    }
+
+    public Address[] Signatories => GetSignatories(primaryGroup);
+
+    public uint Quorum => GetQuorum(primaryGroup);
+
+    private void VerifySignatures(byte[] signatures, string authorizationChallenge)
+    {
+        string[] sigs = this.Serializer.ToArray<string>(signatures);
+        Assert(ECRecover.TryVerifySignatures(sigs, System.Text.Encoding.ASCII.GetBytes(authorizationChallenge), this.Signatories, out Address[] verifiedSignatures) && verifiedSignatures.Length >= this.Quorum,
+            $"Please provide {this.Quorum} valid signatures for '{authorizationChallenge}'.");
+    }
+
+    public Address[] GetSignatories(string group)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        return this.State.GetArray<Address>($"Signatories:{group}");
+    }
+
+    public uint GetQuorum(string group)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        return this.State.GetUInt32($"Quorum:{group}");
+    }
+
+    public void AddSignatory(byte[] signatures, string group, Address address, uint newSize, uint newQuorum)
+    {
+        Assert(!string.IsNullOrEmpty(group));
+        Assert(newSize >= newQuorum, "The number of signatories can't be less than the quorum.");
+
+        Address[] signatories = this.GetSignatories(group);
+        for (int i = 0; i < signatories.Length; i++)
+            Assert(signatories[i] != address, "The signatory already exists.");
```

Comment:
Adding, removing or checking existence of signatories cause iterative codes and resizing array costs. Alternatively we can index signatories in state like 

"Signatories:{signature}" => group

After that you can quickly verify it is existence in dictionary 

Another key pair alternative would be 

"Signatories:{group}:{signature}" => bool(true)

@rowandh  wdty?

Reply:
That would shift the problem of being able to obtain a list of all signatories.

Reply:
Proably "Signatories:{signature}" => group wouldn't work because same address can be added to different groups so second  solution is only solution.

Reply:
@YakupIpek , given your approach how would you list all the signatories of a group?